JSON-Grammar-spec.md

```diff
@@ -0,0 +1,794 @@
+
+```typescript
+interface Grammar {
+  start: string;                              // Entry point rule name
+  cst: Record<string, GrammarNode>;          // Concrete syntax rules
+  ast?: Record<string, AstNodeExpression>;   // AST transformation rules
+}
+```
+
+### Basic Example
+
+```json
+{
+  "start": "Value",
+  "cst": {
+    "Value": {"r": "Number"},
+    "Number": {"t": "/\\d+/"}
+  },
+  "ast": {
+    "Number": ["num", ["$", "/raw"]]
+  }
+}
+```
+
+### Structure Components
+
+#### Start Symbol
+- **`start`**: String specifying the root grammar rule name
+- Must reference a rule defined in the `cst` section
+- Determines the entry point for parsing
+
+#### CST Rules
+- **`cst`**: Object mapping rule names to grammar node definitions
+- Contains all named grammar rules that can be referenced
+- Supports all five grammar node types (Reference, Terminal, Production, Union, List)
+
+#### AST Transformations
+- **`ast`**: Optional object defining custom AST generation rules
+- Maps rule names to JSON expressions for AST transformation
+- Overrides default AST generation behavior
+
+## Grammar Node Types
+
+The JSON grammar specification supports five fundamental node types for defining parsing rules. Each type has a specific JSON representation with both full interface and shorthand syntax options.
+
+### 1. RefNode (Reference Node)
+
+References a named rule defined elsewhere in the grammar.
+
+#### Interface
+```typescript
+type RefNode<Name extends string = string> = {
+  r: Name;
+};
+```
+
+#### Syntax
+```json
+{"r": "RuleName"}
+```
+
+#### Examples
+```json
+{
+  "start": "Program",
+  "cst": {
+    "Program": {"r": "Statement"},
+    "Statement": "return;"
+  }
+}
+```
+
+### 2. TerminalNode (Terminal Node)
+
+Matches literal strings, regular expressions, or arrays of alternatives. Terminal nodes are leaf nodes in the parse tree.
+
+#### Interface
+```typescript
+interface TerminalNode {
+  type?: string;                           // Type name (default: "Text")
+  t: RegExp | string | '' | string[];      // Pattern(s) to match
+  repeat?: '*' | '+';                      // Repetition (only for string arrays)
+  sample?: string;                         // Sample text for generation
+  ast?: AstNodeExpression;                 // AST transformation
+}
+
+// Shorthand: string, RegExp, or empty string
+type TerminalNodeShorthand = RegExp | string | '';
+```
+
+#### Syntax Options
+
+**String Literal:**
+```json
+"hello"
+```
+
+**Regular Expression:**
+```json
+{"t": "/[a-z]+/"}
+```
+
+Note: Regular expressions in JSON must be represented as objects with a `t` property containing the regex pattern as a string.
+
+**Array of Alternatives:**
+```json
+{"t": ["true", "false"]}
+```
+
+**With Repetition:**
+```json
+{"t": [" ", "\t", "\n"], "repeat": "*"}
+```
+
+**Full Terminal Node:**
+```json
+{
+  "t": "/\\d+/",
+  "type": "Number",
+  "sample": "123",
+  "ast": ["num", ["$", "/raw"]]
+}
+```
```

Comment:
Convert all these code blocks to `js`. And add comments with examples which show what this grammar node matches.